specs/Proposals/0040-split-states.md

```diff
@@ -0,0 +1,381 @@
+We cannot simply split transactions and receipts to the new shards and process updates on each new shard separately.
+If we do so, since each shard processes transactions and receipts with their own gas limits, some receipts may be delayed in the new states but not in the current states, or the other way around.
+That will lead to inconsistencies between the orderings by which transactions and receipts are applied to the current and new states.
+
+For example, for simplicity, assume there is only one shard A in epoch T-1 and there will be two shards B and C in epoch T.
+ To process a block in epoch T-1, shard A needs to process receipts 0, 1, .., 99 while in the new sharding assignments receipts 0, 2, …, 98 belong to shard B and receipts 1, 3, …, 99 belong to shard C.
+Assume in shard A, the gas limit is hit after receipt 89 is processed, so receipts 90 to 99 are delayed.
+To achieve the same processing result, shard B must process receipt 0, 2, …, 88 and delay 90, 92, ..., 98 and shard C must process receipt 1, 3, ..., 89 and delay receipts 91, 93, …, 99.
+However, shard B and C have their own gas limits and which receipts will be processed and delayed cannot be guaranteed.
+
+Whether a receipt is processed in a block or delayed can affect the execution result of this receipt because transactions are charged and local receipts are processed before delayed receipts are processed.
+For example, let’s assume Alice’s account has 0N now and Bob sends a transaction T1 to transfer 5N to Alice.
+The transaction has been converted to a receipt R at block i-1 and sent to Alice's shard at block i.
+Let's say Alice signs another transaction T2 to send 1N to Charlie and that transaction is included in block i+1.
+Whether transaction T2 succeeds depends on whether receipt R is processed or delayed in block i.
+If R is processed in block i, Alice’s account will have 5N before block i+1 and T2 will succeed while if R is delayed in block i, Alice’s account will have 0N and T2 will be declined.
+
+Therefore, the validators must still process transactions and receipts based on the current sharding assignment.
+After the processing is finished, they can take the generated state changes to apply to the new states.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+## Protocol-Level Shard Representation
+### `ShardLayout`
+```rust
+pub enum ShardLayout {
+    V0(ShardLayoutV0),
+    V1(ShardLayoutV1),
+}
+```
+ShardLayout is a versioned struct that contains all information needed to decide which accounts belong to which shards. Note that `ShardLayout` only contains information at the protocol level, so it uses `ShardOrd` instead of `ShardId`. 
+
+The API contains the following two functions.
+#### `parent_shards`
+```rust
+pub fn parent_shards(&self) -> Vec<ShardOrd>
```

Comment:
We no longer use `ShardOrd` right?

Reply:
Yes, sorry I haven't been keeping this NEP updated enough since there are too many changes. Just updated that.